proxystore/store/dim/margo.py

```diff
@@ -0,0 +1,411 @@
+                the cache is disabled. The cache is local to the Python
+                process (default: 16).
+            stats (bool): collect stats on store operations (default: False).
+        """
+        global server_process
+
+        # raise error if modules not properly loaded
+        if pymargo_import_error is not None:  # pragma: no cover
+            raise pymargo_import_error
+
+        self.protocol = protocol
+
+        self.host = get_ip_address(interface)
+        self.port = port
+
+        self.addr = f'{self.protocol}://{self.host}:{port}'
+
+        if server_process is None:
+            server_process = Process(target=self._start_server)
+            server_process.start()
+
+        # allocate some time to start the server process
+        sleep(0.2)
+
+        # start client
+        self.engine = Engine(self.protocol, mode=pymargo.client)
+
+        self._rpcs = {
+            'set': self.engine.register('set_bytes'),
+            'get': self.engine.register('get_bytes'),
+            'exists': self.engine.register('exists'),
+            'evict': self.engine.register('evict'),
+        }
+
+        super().__init__(
+            name,
+            cache_size=cache_size,
+            stats=stats,
+            kwargs={
+                'interface': interface,
+                'port': self.port,
+                'protocol': self.protocol,
+            },
+        )
+
+    def _start_server(self) -> None:
+        """Launch the local Margo server (Peer) process."""
+        logger.info(f'starting server {self.addr}')
+        server_engine = Engine(self.addr)
+
+        logger.info(f'Server running at address {str(server_engine.addr())}')
+
+        server_engine.on_finalize(when_finalize)
+        server_engine.enable_remote_shutdown()
+
+        # create server
+        MargoServer(server_engine)
+
+        server_engine.wait_for_finalize()
+
+    def create_key(self, obj: Any) -> MargoStoreKey:
+        return MargoStoreKey(
+            margo_key=utils.create_key(obj),
+            obj_size=len(obj),
+            peer=self.addr,
+        )
+
+    def evict(self, key: MargoStoreKey) -> None:
+        logger.debug(f'Client issuing an evict request on key {key}')
+        self.call_rpc_on(
+            self.engine,
+            key.peer,
+            self._rpcs['evict'],
+            '',
+            key.margo_key,
+            0,
+        )
+
+        self._cache.evict(key)
+
+    def exists(self, key: MargoStoreKey) -> bool:
+        logger.debug(f'Client issuing an exists request on key {key}')
+        buff = bytearray(1)  # equivalent to malloc
+
+        blk = self.engine.create_bulk(buff, bulk.read_write)
+
+        self.call_rpc_on(
+            self.engine,
+            key.peer,
+            self._rpcs['exists'],
+            blk,
+            key.margo_key,
+            len(buff),
+        )
+        return bool(int(bytes(buff).decode(ENCODING)))
```

Comment:
There's some inconsistencies with where `status` is checked (just in `call_on_rpc` or in the calling method). Also if `Success` contains an error, should it be raised? If the errors are unexpected they should be raised I think.

Reply:
agreed, but wanted to mimic the errors of other stores? I think for redis no error is raised if the key doesn't exist (iirc) ?

Reply:
Right, I'm not saying to change the results the user sees, just to be consistent about who handles unexpected errors. Does the caller of `call_rpc_on` raise exceptions or does `call_rpc_on` handle exceptions itself and asserts that if it returns the operation was successful. Right now it's half and half, and it's not possible for the caller to handle an expected error (I know there are no expected errors at the moment but that could change).

Reply:
so the reason `get_bytes` verifies the success is because the other stores return `None` when the key doesn't exist. I wanted the errors to be logged despite the fact that the other rpcs don't really "care", so `call_rpc_on` makes sure to log the error. I can make each RPC log (or raise?) the error independently, if that's cleaner.

Reply:
It seems both the UCX and Websockets implementations let the caller deal with `Status` so I think it makes most sense to stick with that paradigm for Margo as well.

In the Margo case, it seems the server always logs its exceptions as well, so I think `MargoStore` can be selective about what it needs to log vs raise vs ignore.

Reply:
ok, for now i just removed the section in call_rpc_on and haven't added any error handling in the client functions. I suppose we'll add it later?

Reply:
I mentioned below, but we should still at least raise an exception we get since MargoServer catches all expcetions.